### PR TITLE
Generate targets in metaproj

### LINKED ISSUE #1275: Solution targets need to be dynamic or at least extensible
There's currently a fixed set of targets emitted to the sln metaproj:

https://github.com/Microsoft/msbuild/blob/2db4d4b6821ff65987b2843d29cc5f02c4db59f5/src/XMakeBuildEngine/Construction/Solution/SolutionProjectGenerator.cs#L765-L769 (edited)

We will be needing other targets such as "pack" and "test" to work on solutions. 

@jeffkl suggested "It might be as simple as taking the target you specified at the command-line and injecting it into the generated project"

### PR DESCRIPTION
**Pull request by jeffkl**
This allows users to access any target when building a solution.  The target may or may not exist but this will create the corresponding targets in the `.metaproj`.

I had to pass along the target names from the build request all the way down to the solution generator via ReadOnlyCollections.

Closes #1275

### CODE REVIEW

File: src/XMakeBuildEngine/BackEnd/Shared/BuildRequestConfiguration.cs [hunk 1]
```diff
@@ -140,6 +142,8 @@ internal class BuildRequestConfiguration : IEquatable<BuildRequestConfiguration>
 
         #endregion
 
+        internal IReadOnlyCollection<string> TargetNames { get; }
```
Comment (AndyGerlicher) [inline src/XMakeBuildEngine/BackEnd/Shared/BuildRequestConfiguration.cs]: This could probably use a doc comment.

**Comment by jeffkl** [comment]
@cdmihai Please review

File: src/XMakeBuildEngine/BackEnd/Shared/BuildRequestConfiguration.cs [hunk 2]
```diff
@@ -7,6 +7,7 @@
 
 using System;
 using System.Collections.Generic;
+using System.Collections.ObjectModel;
```
Comment (cdmihai) [inline src/XMakeBuildEngine/BackEnd/Shared/BuildRequestConfiguration.cs]: Who needs this namespace?

File: src/XMakeBuildEngine/Construction/Solution/SolutionProjectGenerator.cs [hunk 3]
```diff
@@ -104,6 +112,11 @@ internal class SolutionProjectGenerator
         private ILoggingService _loggingService;
 
         /// <summary>
+        /// The list of targets specified to use.
+        /// </summary>
+        private readonly IReadOnlyCollection<string> _targetNames = new Collection<string>();
```
Comment (cdmihai) [inline src/XMakeBuildEngine/Construction/Solution/SolutionProjectGenerator.cs]: Ah, not all paths write to this fields. Ignore.

---
In reply to: [93343487](https://github.com/Microsoft/msbuild/pull/1497#discussion_r93343487) [](ancestors = 93343487)

**Pull request merged**